### PR TITLE
Added smart merging for loaders

### PR DESCRIPTION
**Pull request by GreenGremlin**
This PR adds logic to avoid merging loaders with matching test values.

### CODE REVIEW

**Comment by bebraw** [comment]
Can you add a test for this?

I wonder if this is a good default behavior, though. I know it could be a valid idea to have the same test multiple times in certain scenarios but maybe it's a configuration design issue then? If so, we might want to print out a warning that merging happened based on the rule so that the user is aware of the behavior as it might not be obvious.

**Comment by GreenGremlin** [comment]
It might make sense to merge the loader(s) values into a single entry. I'll have to give it some more thought.

**Comment by bebraw** [comment]
Ok, cool.

We can also expand the API if needed (another function?). I'm not a big fan of flags.

**Comment by GreenGremlin** [comment]
I made the merging a bit smarter and added tests. I also added eslint and a build task using Babel, I hope that's ok.

**Comment by GreenGremlin** [comment]
I'm trying to find information on how important it is to support Node 0.12, now that Node 4 is stable. This change currently requires Node 4.

**Comment by bebraw** [comment]
@GreenGremlin Ok, maybe we could consume [find](https://lodash.com/docs#find) from lodash (same idea as for other lodash deps).

**Comment by GreenGremlin** [comment]
Good call. Done.

File: test.js [hunk 1]
```diff
@@ -44,33 +46,132 @@ describe('Merge', function() {
...
-      foo: ['c'],
+      foo: ['c']
     };
 
     assert.deepEqual(merge(a, b, c), {
       foo: ['a', 'b', 'c']
     });
   });
+
+  it('should overwride loader string values', function () {
```
Comment (GreenGremlin) [inline test.js]: This should mimic merging as if the test was the key. We always override when given string values (loaders.loader), we merge arrays when given array values (loaders.loaders). Query params are always taken from the final value.

Comment (bebraw) [inline test.js]: Yeah. That's a good solution! Maybe this behavior should be mentioned at the README. Apart from that I'm fine with this as it keeps the API clean.

File: .travis.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
 language: node_js
 node_js:
-  - "0.10"
-  - "0.12"
+  - "0.12.9"
```
Comment (bebraw) [inline .travis.yml]: Any thoughts on *0.10*? Is there any reason why not to support it as well? I know it's very old so dropping support may be justifiable.

**Comment by GreenGremlin** [comment]
Good point, 0.10 isn't set to EOL until next year sometime.

**Comment by bebraw** [comment]
@GreenGremlin Anything else on mind? We might not need *babel-polyfill* anymore but I can look into that after merge.

Btw I don't mind adding you to the project as a contributor.

**Comment by GreenGremlin** [comment]
I missed that, thanks.

**Comment by bebraw** [comment]
`0.4.0` is out. Thanks for contributing!

**Comment by laszlokorte** [comment]
This broke my stylus file loading.

My config looks like:

dev.config.babel.js
```es6
import merge from 'webpack-merge';

import commonConfig from './common.config.babel.js';

module.exports = merge(commonConfig, {
  debug: true,
  devtool: "cheap-module-inline-source-map",
  profile: false,

  watch: true,
  watchOptions: {
    aggregateTimeout: 300,
    poll: 1000,
  },

  module: {
    loaders: [
      {
        test: /\.css$/,
        loader: "style-loader!css-loader",
      },

      {
        test: /\.styl$/,
        loader: "style-loader!css-loader!stylus-loader?resolve url",
      },
    ],
  },

  plugins: [
  ],

  devServer: {
    contentBase: "./app/static",
    port: 3000,

    historyApiFallback: true,

    colors: true,
    stats: 'normal',
  },
});
```

common.config.babel.js
```es6
import webpack from 'webpack';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import ExtractTextPlugin from 'extract-text-webpack-plugin';
import CopyWebpackPlugin from 'copy-webpack-plugin';
import jeet from 'jeet';
import rupture from 'rupture';
import nib from 'nib';

const vendorModules = /(node_modules|bower_components)/;

const htmlMinifyOptions = process.env.NODE_ENV === 'production' ? {
  removeComments: true,
  removeCommentsFromCDATA: true,
  collapseWhitespace: true,
  conservativeCollapse: false,
  collapseBooleanAttributes: true,
  removeAttributeQuotes: true,
  removeRedundantAttributes: true,
  preventAttributesEscaping: true,
  useShortDoctype: true,
  removeEmptyAttributes: true,
  removeScriptTypeAttributes: true,
  removeStyleLinkTypeAttributes: true,
} : false;

//export default
module.exports = {
  target: "web",
  entry: {
    app: "./app/index.js",
    experiment: "./app/experiment.js",
    debug: "./app/debug.js",
    vendor: require("../app/vendor.js"),
  },

  output: {
    path: "./build",
    filename: "[name]-[hash].js",
    pathinfo: true,
    publicPath: "",
  },

  module: {
    preLoaders: [
      {test: /\.js$/, loader: "eslint-loader", exclude: vendorModules},
    ],
    loaders: [
      // {
      //   test: /\.(json)$/,
      //   loader: "file?name=[path][name].[ext]&context=./app/static",
      // },
      {
        test: /\.(png|jpg|svg|ttf|eot|woff|woff2)$/,
        loader: 'url?name=[path][name].[ext]',
      },
      {
        test: /\.js$/,
        exclude: vendorModules,
        loader: "babel",
      },
    ],
  },

  stylus: {
    use: [jeet(), rupture(), nib()],
  },

  plugins: [
    new webpack.optimize.CommonsChunkPlugin(
      'vendor', 'vendor-[hash].js', Infinity
    ),
    new ExtractTextPlugin('[name]-[contenthash].css', {allChunks: true}),
    new HtmlWebpackPlugin({
      title: 'KV Editor',
      minify: htmlMinifyOptions,
      chunks: ['app', 'vendor'],
      template: './app/index.html',
      favicon: './app/favicon.ico',
    }),
    new HtmlWebpackPlugin({
      title: 'Experiment',
      minify: htmlMinifyOptions,
      chunks: ['experiment', 'vendor'],
      template: './app/index.html',
      filename: 'experiment.html',
      favicon: './app/favicon.ico',
    }),
    new HtmlWebpackPlugin({
      title: 'Debug',
      minify: htmlMinifyOptions,
      chunks: ['debug', 'vendor'],
      template: './app/index.html',
      filename: 'debug.html',
      favicon: './app/favicon.ico',
    }),
    new CopyWebpackPlugin([
      {from: 'app/static'},
    ]),
  ],
};
```

As error message I get
```
ERROR in ./app/components/main/view.styl
Module parse failed: /[...]/app/components/main/view.styl Line 1: Unexpected token .
You may need an appropriate loader to handle this file type.
```

**Comment by bebraw** [comment]
@laszlokorte I moved the smart behavior behind a function of its own at 0.5.0, publisher today. Can you try against that?

**Comment by laszlokorte** [comment]
0.5.0 is the version which is not working for me. Which merge function should I use? `merge()` oder `merge.smart()`

**Comment by bebraw** [comment]
On Saturday, 26 December 2015, Laszlo Korte <notifications@github.com>
wrote:

> 0.5.0 is the version which is not working for me. Which merge function
> should I use? merge() oder merge.smart()
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/survivejs/webpack-merge/pull/5#issuecomment-167337910>
> .
>

merge should be the same as in 0.3.x.

I'll set up a little test case based on your example and study this further.

**Comment by laszlokorte** [comment]
FYI ^0.3.0 has been working for the last few weeks. on 23rd of dec I upgraded to 0.4.0 which has been working as well. Today I upgraded to 0.5.0 and the error occurred. Now I downgraded to 0.4.0 again to get it working again.

**Comment by bebraw** [comment]
@laszlokorte Yeah, I figured it out. I was missing a `bind` and tests didn't catch it. `0.5.1` contains the fix. Thanks for letting me know and my apologies for the inconvenience. :+1:

**Comment by laszlokorte** [comment]
Thank you very much for fixing it that fast!

**Pull request merged**